doc/dependency-resolution.md

```diff
@@ -0,0 +1,319 @@
+has a parent of B, we should track this by adding B into A's links in the
+'parent' group. For the most part, we already do this.
+
+To specify the fields that a content item depends on, we should do so in the
+content schemas. This set of fields varies by use case, which relates to the
+name of the group. For example, the 'parent' group requires a 'title' and a
+'base_path' and so it is proposed to capture this data on a group basis to apply
+generally across all formats. If required, we can add a finer level of control
+over this in the future – perhaps to specialise it on a group/format basis.
+
+Additionally, the recursive behaviour for dependency tracking should be captured
+in the schemas. We'd capture that the 'parent' group should recurse in order to
+track the full ancestry of content items as its dependencies. This differs for
+the 'related' group, which would not need to recurse. Therefore, I propose that
+we have a simple boolean value that represents whether to recurse or not. Again,
+we can add a finer level of control for this later if required.
+
+Finally, we should specify the default behaviour when custom configuration is
+not specified. This should also go into the schemas and exactly match the way in
+which link expansion works currently. i.e. it should not recurse and it should
+contain the set of fields that are currently expanded (title, base_path, etc).
+
+Note: Bidirectional dependencies is documented separately later on.
+
+- Detecting when dependencies change
+
+The Publishing API can receive interactions that affect dependencies. When we
+receive one of these interactions for a given content_id, we should explicitly
+check its dependencies and send updates downstream. Because the rules for how
+dependencies are tracked is captured in the content schemas, it is proposed that
+Publishing API be granted visibility over the content schemas in production so
+that it can apply these rules when detecting changes. We had already planned to
+do this for the purpose of validating content against the schemas in the
+Publishing API.
+
+Once it has visibility over the dependency rules for each of the groups, it
+should run a query against the database to detect content items that need to be
+updated as a result of the incoming request. It will do this using Postgres's
+WITH RECURSIVE functionality. This was explored in the following branch to
+demonstrate that it is feasible to detect dependent nodes in this way:
+
+[https://github.com/alphagov/publishing-api/tree/dependency-transitive-closure](https://github.com/alphagov/publishing-api/tree/dependency-transitive-closure)
+
+We should make it so that this query only follows edges in the dependency graph
+for groups which contain any of the fields that have changed on the content
+item. We can determine which fields have changed using Rails's
+[ActiveModel::Dirty](http://api.rubyonrails.org/classes/ActiveModel/Dirty.html)
```

Comment:
This will only track individual ActiveModel fields, it won't be able to tell that e.g. `details.start_date` has changed but not `details.end_date`.